Document Optional annotation feature (#49693)
openshift/openshift-docs: modules/sbo-setting-annotations-mapping-optional.adoc

:_content-type: PROCEDURE
[id="sbo-setting-annotations-mapping-optional_{context}"]
= Setting annotations mapping to be optional

> You can have optional fields in the annotations. For example, a link to the credentials might not be there if the service endpoint does not require authentication. In such cases, a field might not exist in the target path of the annotations. As a result, {servicebinding-title} generates an error, by default. 

Author: @baijum @pmacik WDYT? Please suggest. Thanks!

> You can have optional fields in the annotations. For example, a link to the credentials might not be there if the service endpoint does not require authentication. In such cases, a field might not exist in the target path of the annotations. As a result, {servicebinding-title} generates an error, by default. 

Reviewer: yes, `path` seems better... also instead of `might not be there` I'd suggest to use `might not be present`

> You can have optional fields in the annotations. For example, a link to the credentials might not be there if the service endpoint does not require authentication. In such cases, a field might not exist in the target path of the annotations. As a result, {servicebinding-title} generates an error, by default. 

Author: Done.

> You can have optional fields in the annotations. For example, a link to the credentials might not be there if the service endpoint does not require authentication. In such cases, a field might not exist in the target path of the annotations. As a result, {servicebinding-title} generates an error, by default. 

Author: Done.

> As a service provider, to indicate whether you require annotations mapping, you can set a value to the `optional` flag in your annotations when enabling services. {servicebinding-title} provides annotations mapping only if the target path is available. When the target path is not available, the {servicebinding-title} skips providing the annotations mapping. 

Reviewer 2: I would say "skips the optional mapping and continues with the projection of the existing mappings without throwing any errors" something like that. The wording "skips providing the annotations mapping" is a bit ambiguous in my opinion.

Author: Thanks, @dperaza4dustbit! This looks much better :)

Author: Done.

> As a service provider, to indicate whether you require annotations mapping, you can set a value to the `optional` flag in your annotations when enabling services. {servicebinding-title} provides annotations mapping only if the target path is available. When the target path is not available, the {servicebinding-title} skips the optional mapping and continues with the projection of the existing mappings without throwing any errors.

Reviewer 3: ```suggestion
As a service provider, to indicate whether you require annotations mapping, you can set a value for the `optional` flag in your annotations when enabling services. {servicebinding-title} provides annotations mapping only if the target path is available. When the target path is not available, the {servicebinding-title} skips the optional mapping and continues with the projection of the existing mappings without throwing any errors.
```

Author: Done

.Procedure

> * To enable {servicebinding-title} consider a new field in the annotations, provide the mapping by setting the `optional` flag value to `true`:

Reviewer: ```suggestion
* To make a new field in the annotations optional set the `optional` flag value to `true`:
```

Author: @pmacik Thanks! I like this. Just a small change from my side:
"To make a field in the annotations optional, set the `optional` flag value to `true`:"

Reason: It is for all fields and not just new fields.

Author: Done

+
.Example
[source,yaml]
----
apiVersion: apps.example.org/v1beta1
kind: Database
metadata:
  name: my-db
  namespace: my-petclinic
  annotations:
    service.binding/username: path={.spec.name},optional=true
    ...
----

[NOTE]
====

> * If you set the `optional` flag value to `false` and when the {servicebinding-title} is unable to find the target path, the Operator fails the annotations mapping.

Reviewer 3: ```suggestion
* If you set the `optional` flag value to `false` and the {servicebinding-title} is unable to find the target path, the Operator fails the annotations mapping.
```

Author: Done
